### ISSUE TITLE
Nodeport service cannot be accessed from outside cluster if kube-proxy mode is IPVS and cni plugin is calico.

### ISSUE THREAD

**Issue by paulsubrata55** [kubernetes/kubernetes#73300]
**What happened**:
Started a cluster using following config:
```
apiVersion: kubeadm.k8s.io/v1beta1
kind: InitConfiguration
localAPIEndpoint:
  advertiseAddress: "192.168.56.101"
  bindPort: 6443
---
kind: ClusterConfiguration
apiVersion: kubeadm.k8s.io/v1beta1
kubernetesVersion: v1.13.1
networking:
  podSubnet: "192.168.0.0/16"
apiServer:
  extraArgs:
    feature-gates: SCTPSupport=true
---
apiVersion: kubeproxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
mode: ipvs
```
After that I installed calico for the CNI. Then created a deployment followed by a Nodeport service.
In my cluster I have 2 nodes with 192,168.56.101 as master and 192,168.56.102 as worker. All of my pods from the above deployment are scheduled on worker node. Now if I use 192,168.56.102:nodeport to reach the service, it works fine. But when I use 192,168.56.101:nodeport for the same, it does not. For both of the cases I used another system which is not part of the cluster.

**What you expected to happen**:
A nodeport service can be contacted from any node using nodeip:nodeport

**How to reproduce it (as minimally and precisely as possible)**:
To reproduce the issue:
1. Use 2 nodes (1 master, 1 worker).
2. Init the cluster with ipvs enabled.
3. Allow scheduling of pods only on worker node.
4. Apply a deployment and start a nodeport service.
5. Try to connect to the nodeport service using master's nodeip from a system outside the cluster. 

**Anything else we need to know?**:
My observation is : when the nodeip:nodeport packet reached to the master the Ipvs rule got applied and the dst ip is changed to a pod ip by the ipvs, but the src ip remains intact and the packet is sent to the other node since the pod is on that node which is not a correct behavior.
The expected behavior should be an iptable postrouting masquerade rule must be applied on that packet before sending to the other node for SNAT which is not happening.

**Environment**:
- Kubernetes version v1.13.0:
- Cloud provider or hardware configuration: Virtual box VMs
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`): 4.4.0-131-generic
- Install tools: Calio v3.3.2
- Others: IPVS enabled

/sig network
/area kube-proxy

**Comment by paulsubrata55** [comment]
area/ipvs

**Comment by paulsubrata55** [comment]
@rongyi I have set the `clusterCIDR `, still not working. I can see the ipset in my system
```
Name: KUBE-NODE-PORT-SCTP
Type: bitmap:port
Revision: 3
Header: range 0-65535
Size in memory: 8300
References: 1
Members:
30608
```
but the corresponding iptables rule to match this is missing. (`KUBE-SVC-IPSETS` chain is  not present).

**Comment by rongyi** [comment]
@paulsubrata55 after adding clusterCIDR option we solved this problem. can you dump your kube-proxy configmap to here?

**Comment by paulsubrata55** [comment]
I found the problem is with the sctp traffic. For tcp/udp traffic the ipvs node port works fine.
Another observation is that ipset bitmap:port does not support SCTP. But `KUBE-NODE-PORT-SCTP` is using bitmap:port

**Comment by paulsubrata55** [comment]
To fix the issue for sctp traffic I have done a simple patch on ipvs/proxier.go to make things work in my system. Using `hash IP:Port` for ipset `KUBE-NODE-PORT-SCTP` rule instead of `bitmap:port` since `hash ip:port` supports sctp also.

Pasting the diff below. Please review:
```
diff --git a/pkg/proxy/ipvs/proxier.go b/pkg/proxy/ipvs/proxier.go
index 9ca7e05..dbce0b7 100644
--- a/pkg/proxy/ipvs/proxier.go
+++ b/pkg/proxy/ipvs/proxier.go
@@ -127,8 +127,8 @@ var ipsetInfo = []struct {
 	{kubeNodePortLocalSetTCP, utilipset.BitmapPort, kubeNodePortLocalSetTCPComment},
 	{kubeNodePortSetUDP, utilipset.BitmapPort, kubeNodePortSetUDPComment},
 	{kubeNodePortLocalSetUDP, utilipset.BitmapPort, kubeNodePortLocalSetUDPComment},
-	{kubeNodePortSetSCTP, utilipset.BitmapPort, kubeNodePortSetSCTPComment},
-	{kubeNodePortLocalSetSCTP, utilipset.BitmapPort, kubeNodePortLocalSetSCTPComment},
+	{kubeNodePortSetSCTP, utilipset.HashIPPort, kubeNodePortSetSCTPComment},
+	{kubeNodePortLocalSetSCTP, utilipset.HashIPPort, kubeNodePortLocalSetSCTPComment},
 }
 
 // ipsetWithIptablesChain is the ipsets list with iptables source chain and the chain jump to
@@ -153,8 +153,8 @@ var ipsetWithIptablesChain = []struct {
 	{kubeNodePortSetTCP, string(KubeNodePortChain), string(KubeMarkMasqChain), "dst", "tcp"},
 	{kubeNodePortLocalSetUDP, string(KubeNodePortChain), "RETURN", "dst", "udp"},
 	{kubeNodePortSetUDP, string(KubeNodePortChain), string(KubeMarkMasqChain), "dst", "udp"},
-	{kubeNodePortSetSCTP, string(kubeServicesChain), string(KubeNodePortChain), "dst", "sctp"},
-	{kubeNodePortLocalSetSCTP, string(KubeNodePortChain), "RETURN", "dst", "sctp"},
+	{kubeNodePortSetSCTP, string(KubeNodePortChain), string(KubeMarkMasqChain), "dst,dst", "sctp"},
+	{kubeNodePortLocalSetSCTP, string(KubeNodePortChain), "RETURN", "dst,dst", "sctp"},
 }
 
 // In IPVS proxy mode, the following flags need to be set
@@ -1085,20 +1085,32 @@ func (proxier *Proxier) syncProxyRules() {
 
 			// Nodeports need SNAT, unless they're local.
 			// ipset call
-			entry = &utilipset.Entry{
-				// No need to provide ip info
-				Port:     svcInfo.NodePort,
-				Protocol: protocol,
-				SetType:  utilipset.BitmapPort,
-			}
 			var nodePortSet *IPSet
 			switch protocol {
 			case "tcp":
 				nodePortSet = proxier.ipsetList[kubeNodePortSetTCP]
+				entry = &utilipset.Entry{
+					// No need to provide ip info
+					Port:     svcInfo.NodePort,
+					Protocol: protocol,
+					SetType:  utilipset.BitmapPort,
+				}
 			case "udp":
 				nodePortSet = proxier.ipsetList[kubeNodePortSetUDP]
+				entry = &utilipset.Entry{
+					// No need to provide ip info
+					Port:     svcInfo.NodePort,
+					Protocol: protocol,
+					SetType:  utilipset.BitmapPort,
+				}
 			case "sctp":
 				nodePortSet = proxier.ipsetList[kubeNodePortSetSCTP]
+				entry = &utilipset.Entry{
+					IP:       proxier.nodeIP.String(),
+					Port:     svcInfo.NodePort,
+					Protocol: protocol,
+					SetType:  utilipset.HashIPPort,
+				}
 			default:
 				// It should never hit
 				klog.Errorf("Unsupported protocol type: %s", protocol)

```

**Comment by paulsubrata55** [comment]
@m1093782566 Can you please give a comment on this.

**Comment by MrHohn** [comment]
Re-assigning as the fix (https://github.com/kubernetes/kubernetes/pull/74341) is pending on @m1093782566's approval.
/assign @m1093782566
/unassign

(Removing `triage/unresolved` as this seems to be a real bug.)
/remove-triage unresolved
